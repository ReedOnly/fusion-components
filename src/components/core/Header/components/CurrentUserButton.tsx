import React from 'react';
import { useCurrentUser, useFusionContext } from '@equinor/fusion';
import { useIcon, IconProps, MenuSection, MenuItemType, Menu, usePopoverRef } from 'index';

const CurrentUserIcon = (props: IconProps) => {
    const iconFactory = useIcon(
        <path
            fillRule="evenodd"
            clipRule="evenodd"
            d="M12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2ZM12 5C13.66 5 15 6.34 15 8C15 9.66 13.66 11 12 11C10.34 11 9 9.66 9 8C9 6.34 10.34 5 12 5ZM6 15.98C7.29 17.92 9.5 19.2 12 19.2C14.5 19.2 16.71 17.92 18 15.98C17.97 13.99 13.99 12.9 12 12.9C10 12.9 6.03 13.99 6 15.98Z"
            fill="currentColor"
        />
    );

    return iconFactory(props);
};

const CurrentUserDropdown: React.FC = () => {
    const { auth } = useFusionContext();
    const currentUser = useCurrentUser();

    if (!currentUser) {
        return null;
    }

    const sections: MenuSection[] = [
        {
            key: currentUser.id,
            title: `${currentUser.givenName} ${currentUser.familyName}`,
            items: [
                {
                    key: 'logout',
                    title: 'Sign out',
                },
            ],
            // TODO: Allow core and apps to add custom buttons like "Preview features" etc.
            // Should be provided through the fusion context, not props
        },
    ];

    const onClick = async (item: MenuItemType) => {
        switch (item.key) {
            case 'logout':
                await auth.container.logoutAsync();
                break;
        }
    };

    return <Menu sections={sections} onClick={onClick} elevation={0} />;
};

const CurrentUserButton: React.FC = () => {
    const currentUser = useCurrentUser();
    const popoverRef = usePopoverRef<HTMLButtonElement>(<CurrentUserDropdown />, {
        placement: 'below',
        justify: 'end',
        centered: true,
        fillWithContent: true,
    });

    if (!currentUser) {
        return null;
    }

    return (
        <button ref={popoverRef}>
            <CurrentUserIcon />
        </button>
    );
};

export default CurrentUserButton;
