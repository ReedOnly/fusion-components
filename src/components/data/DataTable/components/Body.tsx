import React from 'react';
import { DataTableBodyProps } from '../types';
import Row from './Row';
import { getString, getBoolean } from '../utils';

function DataTableBody<T>({
    data,
    columns,
    collapsedColumns,
    rowIdentifier,
    isExpandable,
    expandedComponent,
    onExpand,
    expandedItems,
}: DataTableBodyProps<T>) {
    return (
        <>
            {data.map((item, index) => (
                <Row
                    key={getString(item, rowIdentifier)}
                    item={item}
                    columns={columns}
                    collapsedColumns={collapsedColumns}
                    index={index}
                    isExpandable={isExpandable ? getBoolean(item, isExpandable) : false}
                    isExpanded={expandedItems.findIndex(i => i === item) > -1}
                    expandedComponent={expandedComponent}
                    onExpand={() => onExpand(item, index)}
                />
            ))}
        </>
    );
}

export default DataTableBody;
