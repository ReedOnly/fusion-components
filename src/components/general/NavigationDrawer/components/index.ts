export { default as MenuChild } from "./Child";