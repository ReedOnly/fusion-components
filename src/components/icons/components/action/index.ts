export { default as SearchIcon } from './SearchIcon';
export { default as DoneIcon } from './DoneIcon';
export { default as MinimizeIcon } from './MinimizeIcon';
