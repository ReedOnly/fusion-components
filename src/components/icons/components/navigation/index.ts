export { default as PaginationArrow } from "./PaginationArrow";
export { default as CloseIcon } from "./CloseIcon";