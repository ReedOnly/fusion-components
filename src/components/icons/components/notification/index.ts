export { default as SyncDisabledIcon } from './SyncDisabledIcon';
