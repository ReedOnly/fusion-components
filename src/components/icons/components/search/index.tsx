import * as React from 'react';
import { IconProps, useIcon } from 'index';

const SearchIcon = (props: IconProps) => {
    const iconFactory = useIcon(
        <path
            fillRule="evenodd"
            clipRule="evenodd"
            d="M12.1865 11.4716H12.9765L17.2165 15.7316C17.6265 16.1416 17.6265 16.8116 17.2165 17.2216C16.8065 17.6316 16.1365 17.6316 15.7265 17.2216L11.4765 12.9716V12.1816L11.2065 11.9016C9.80646 13.1016 7.89646 13.7216 5.86646 13.3816C3.08646 12.9116 0.866456 10.5916 0.526456 7.79157C0.00645569 3.56157 3.56646 0.00157288 7.79646 0.521573C10.5965 0.861573 12.9165 3.08157 13.3865 5.86157C13.7265 7.89157 13.1065 9.80157 11.9065 11.2016L12.1865 11.4716ZM2.47646 6.97157C2.47646 9.46157 4.48646 11.4716 6.97646 11.4716C9.46646 11.4716 11.4765 9.46157 11.4765 6.97157C11.4765 4.48157 9.46646 2.47157 6.97646 2.47157C4.48646 2.47157 2.47646 4.48157 2.47646 6.97157Z"
            fill="currentColor"
        />
    );

    return iconFactory(props);
};

export default SearchIcon;
