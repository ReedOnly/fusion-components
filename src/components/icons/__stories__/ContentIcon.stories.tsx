import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { actions } from '@storybook/addon-actions';
import { withKnobs, number, color } from '@storybook/addon-knobs';

import { BlockIcon } from '../components/content';

const eventsFromNames = actions('onClick');

const stories = storiesOf('Icons|Content', module);
stories.addDecorator(withKnobs);

stories.add('Block', () => (
    <BlockIcon
        {...eventsFromNames}
        width={number('width', 24)}
        height={number('height', 24)}
        color={color('color', '#000')}
    />
));
