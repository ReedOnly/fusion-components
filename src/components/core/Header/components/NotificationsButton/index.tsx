import * as React from 'react';
import * as styles from './styles.less';
import classNames from 'classnames';
import { useComponentDisplayClassNames, useNotificationCards } from '@equinor/fusion';
import NotificationCardsPresenter from '../../../NotificationCardsPresenter';
import useIcon, { IconProps } from 'hooks/useIcon';
import styling from 'styles/styling';
import IconButton from 'components/general/IconButton';
import { NotificationsIcon } from 'components/icons/components/notification';
import NotificationsSideSheet from 'components/general/NotificationsSideSheet';
import OverlayPortal from 'components/utils/OverlayPortal';

type NotificationNumberBadgeProps = IconProps & {
    numberOfUnread: number;
};

const NotificationNumberBadge = (props: NotificationNumberBadgeProps) => {
    const iconFactory = useIcon(
        <>
            <circle cx="12" cy="12" r="12" fill={styling.colors.danger} />
            <text x="50%" y="50%" textAnchor="middle" stroke="white" strokeWidth="1px" dy=".3em">
                {props.numberOfUnread}
            </text>
        </>
    );

    return iconFactory(props);
};
const NotificationsButton: React.FC = () => {
    const [showSideSheet, setShowSideSheet] = React.useState<boolean>(false);
    const [hasOpenedSideSheet, setHasOpenedSideSheet] = React.useState<boolean>(false);

    const {
        notificationCards,
        isFetchingRead,
        isFetchingUnRead,
        getReadNotificationCardsAsync,
    } = useNotificationCards();

    const numberOfUnread = React.useMemo(
        () => notificationCards.filter((notification) => !notification.seenByUser).length,
        [notificationCards]
    );
    const badgeClassNames = classNames(
        styles.badgeContainer,
        useComponentDisplayClassNames(styles)
    );

    React.useEffect(() => {
        if (showSideSheet) {
            setHasOpenedSideSheet(true);
        }
    }, [showSideSheet]);

    const openSideSheet = React.useCallback(() => setShowSideSheet(true), [setShowSideSheet]);
    const closeSideSheet = React.useCallback(() => setShowSideSheet(false), [setShowSideSheet]);

    React.useEffect(() => {
        hasOpenedSideSheet && getReadNotificationCardsAsync();
    }, [hasOpenedSideSheet]);

    return (
        <>
            <IconButton onClick={openSideSheet}>
                <div className={styles.notificationsButton}>
                    <NotificationsIcon />
                    {numberOfUnread > 0 && (
                        <div className={badgeClassNames}>
                            <NotificationNumberBadge
                                numberOfUnread={numberOfUnread}
                                width={styling.numericalGrid(3)}
                                height={styling.numericalGrid(3)}
                            />
                        </div>
                    )}
                </div>
            </IconButton>
            <NotificationsSideSheet
                open={showSideSheet}
                onClose={closeSideSheet}
                notifications={notificationCards}
                isFetchingReadNotifications={isFetchingRead}
                isFetchingUnReadNotifications={isFetchingUnRead}
            />
            <OverlayPortal show>
                <NotificationCardsPresenter onShowInList={openSideSheet} />
            </OverlayPortal>
        </>
    );
};

export default NotificationsButton;
