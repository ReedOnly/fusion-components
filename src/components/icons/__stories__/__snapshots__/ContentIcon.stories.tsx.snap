// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icons|Content Add 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <svg
      height={24}
      onClick={[Function]}
      style={
        Object {
          "color": "#000",
        }
      }
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="M18 13H13V18C13 18.55 12.55 19 12 19C11.45 19 11 18.55 11 18V13H6C5.45 13 5 12.55 5 12C5 11.45 5.45 11 6 11H11V6C11 5.45 11.45 5 12 5C12.55 5 13 5.45 13 6V11H18C18.55 11 19 11.45 19 12C19 12.55 18.55 13 18 13Z"
        fill="currentColor"
      />
    </svg>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`Icons|Content Block 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <svg
      height={24}
      onClick={[Function]}
      style={
        Object {
          "color": "#000",
        }
      }
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        clipRule="evenodd"
        d="M10 0C4.48 0 0 4.48 0 10C0 15.52 4.48 20 10 20C15.52 20 20 15.52 20 10C20 4.48 15.52 0 10 0ZM2 10C2 5.58 5.58 2 10 2C11.85 2 13.55 2.63 14.9 3.69L3.69 14.9C2.63 13.55 2 11.85 2 10ZM5.1 16.31C6.45 17.37 8.15 18 10 18C14.42 18 18 14.42 18 10C18 8.15 17.37 6.45 16.31 5.1L5.1 16.31Z"
        fill="currentColor"
        fillRule="evenodd"
      />
    </svg>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`Icons|Content Copy 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <svg
      height={24}
      onClick={[Function]}
      style={
        Object {
          "color": "#000",
        }
      }
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        clipRule="evenodd"
        d="M15.5 1H4.5C3.4 1 2.5 1.9 2.5 3V16C2.5 16.55 2.95 17 3.5 17C4.05 17 4.5 16.55 4.5 16V4C4.5 3.45 4.95 3 5.5 3H15.5C16.05 3 16.5 2.55 16.5 2C16.5 1.45 16.05 1 15.5 1ZM19.5 5H8.5C7.4 5 6.5 5.9 6.5 7V21C6.5 22.1 7.4 23 8.5 23H19.5C20.6 23 21.5 22.1 21.5 21V7C21.5 5.9 20.6 5 19.5 5ZM9.5 21H18.5C19.05 21 19.5 20.55 19.5 20V8C19.5 7.45 19.05 7 18.5 7H9.5C8.95 7 8.5 7.45 8.5 8V20C8.5 20.55 8.95 21 9.5 21Z"
        fill="currentColor"
        fillRule="evenodd"
      />
    </svg>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
